dask/array/routines.py

```diff
@@ -1190,3 +1192,74 @@ def insert(arr, obj, values, axis):
     interleaved = list(interleave([split_arr, split_values]))
     interleaved = [i for i in interleaved if i.nbytes]
     return concatenate(interleaved, axis=axis)
+
+
+def _einsum_kernel(*operands, **kwargs):
+    subscripts = kwargs.pop('subscripts')
+    ncontract_inds = kwargs.pop('ncontract_inds')
+    dtype = kwargs.pop('kernel_dtype')
+    chunk = np.einsum(subscripts, *operands, dtype=dtype, **kwargs)
+
+    # Avoid concatenate=True in atop by adding 1's
+    # for the contracted dimensions
+    return chunk.reshape(chunk.shape + (1,) * ncontract_inds)
+
+
+_einsum_can_optimize = LooseVersion(np.__version__) >= LooseVersion("1.12.0")
+
+
+@wraps(np.einsum)
+def einsum(*operands, **kwargs):
+    casting = kwargs.get('casting', 'safe')
+    dtype = kwargs.get('dtype')
+    optimize = kwargs.get('optimize')
+    order = kwargs.get('order', 'K')
+    einsum_dtype = dtype
+
+    inputs, outputs, ops = parse_einsum_input(operands)
+    subscripts = '->'.join((inputs,outputs))
```

Comment:
there are extra parentheses here

Reply:
Those are necessary (`os.path.join` behaves differently, which is frustrating at times)

```python
>>> '->'.join(('foo', 'qux'))
'foo->qux'
```

Reply:
Oops, I missed the comma.

In that case, please add a space after the comma, per pep8!

Reply:
Done